src/Razor/test/VSCode.FunctionalTest/tests/TestUtil.ts

```diff
@@ -113,50 +116,112 @@ export async function waitForDocumentUpdate(
     return updatedDocument;
 }
 
-export async function dotnetRestore(cwd: string): Promise<void> {
-    return new Promise<void>((resolve, reject) => {
-        const dotnet = cp.spawn('dotnet', ['restore'], { cwd, env: process.env });
+export async function dotnetRestore(directories: string[]): Promise<void> {
+    const promiseList = new Array<Promise<void>>();
 
-        dotnet.stdout.on('data', (data: any) => {
-            console.log(data.toString());
-        });
+    for (const cwd of directories) {
+        const promise = new Promise<void>((resolve, reject) => {
+            const dotnet = cp.spawn('dotnet', ['restore'], { cwd, env: process.env });
 
-        dotnet.stderr.on('err', (error: any) => {
-            console.log(`Error: ${error}`);
-        });
+            // Terminate dotnet restores that are hanging and optimistically continue test execution.
+            let exited = false;
+            setTimeout(() => {
+                if (!exited) {
+                    console.log(`Restore of ${cwd} running for too long. Terminating.`);
+                    dotnet.kill();
+                }
+            }, 180000);
 
-        dotnet.on('close', (exitCode) => {
-            console.log(`Done: ${exitCode}.`);
-            resolve();
-        });
+            dotnet.stdout.on('data', (data: any) => {
+                console.log(data.toString());
+            });
+
+            dotnet.stderr.on('err', (error: any) => {
+                console.log(`Error: ${error}`);
+                reject(error);
+            });
 
-        dotnet.on('error', error => {
-            console.log(`Error: ${error}`);
-            reject(error);
+            dotnet.on('close', (exitCode) => {
+                exited = true;
+                console.log(`Done: ${exitCode}.`);
+                resolve();
+            });
+
+            dotnet.on('error', error => {
+                console.log(`Error: ${error}`);
+                reject(error);
+            });
         });
-    });
+        promiseList.push(promise);
+    }
+
+    // Sometimes restores take a long time before failing. We don't want to block the whole time.
+    // So, let's wait until either all the restores complete or the timeout fires, whichever happens first.
+    const restoresPromise = Promise.all(promiseList);
+    const timeoutPromise = new Promise(r => setTimeout(r, 60000));
+    await Promise.race([restoresPromise, timeoutPromise]);
 }
 
-export async function waitForProjectReady(directory: string) {
+export async function waitForProjectsReady(...directories: string[]) {
+    console.log('Getting test projects ready');
     await removeOldProjectRazorJsons();
-    await cleanBinAndObj(directory);
+
+    if (process.env.norestore !== 'true') {
```

Comment:
Is this for local cases?

Reply:
Yeah. The default is to restore. But we run with `--no-restore` on the CI because we can safely expect it to not be polluted like our local machines.